Modify existentialAbstraction to avoid calls to contains.

diesalbla opened the pull request:
The `existentialAbstraction` method performs a fix-point transitive closure loop, which at each iteration compares a list of "candidate" parameters with a list of "already marked", which is a subset of it.
The test includes both equality (same object) and a call to the `Type.contains` method, which would create a `ContainsCollector` object.

We modify the code, to move quick equality check before doing any call to `Type.contains`. This should avoid creating the `ContainsCollector` object, but would perform two list traversals in each loop.

diesalbla committed:
Modify existentialAbstraction to avoid calls to contains.

The existentialAbstraction performs a transitive closure fix-point
algorithm, which at each iteration compares a list of "candidate"
parameters with a list of "already marked", which is a subset of it.
The test includes both equality (same object) and a call to the
"Type.contains" method, which creates a ContainsCollector object.

We modify the code, to move quick equality check before doing any
call to "Type.contains".

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4132,7 +4132,8 @@ trait Types
       while (tparams1 != tparams0) {
         tparams0 = tparams1
         tparams1 = tparams filter { p =>
```
🤔 should this be `tparams1 = tparams0 filter ...`  (I know this was already there)

smells like an opportunity for redoing this with more fitting data structures?

diesalbla replied:
When I saw this code I found it quite intriguing, and I had doubts as well. But no, it is `tparams1` that it should use.

This is a fixed-point: the `tparams0` is only used to check, by comparison, if the "selected" set `tparams1` has changed. The idea of `tparams1` is to keep a _frontier_, a subset of the input `tparams` that are already known to be directly or indirectly referenced from `tpe`.

There may be an opportunity to redo this with more fitting data structures, yes, and this code seems to may redundant checks. However, I do not know how much relevant this code weighs on overall performance, to use anything more than two lines.

Not that I did not try, though. I will push a different branch with a more elaborate change.

adriaanm replied:
but how about the RHS? Why is it filtering `tparams`, and not `tparams0` each time. `tparams` doesn't change, while `tparams0` does.

diesalbla replied:
The key point is that this is a transitive closure: later iterations of the loop will allow _more_ elements from `tparams` into `tparams1`. If the order of the `tparams` in the final list is not a problem (and if it is an existential it ideally should not be), then we could improve this.

diesalbla replied:
This is the branch where I have made more elaborate changes. The only problem is that `partest` fails.
 https://github.com/scala/scala/compare/2.13.x...diesalbla:existential_abstraction_loops?expand=1

diesalbla replied:
> but how about the RHS? Why is it filtering tparams, and not tparams0 each time. tparams doesn't change, while tparams0 does.

IMHO it would have been more clear if `tparams1` and `tparams0` had been called `existParams` and `existParamsPrev`, respectively.

adriaanm replied:
That makes sense. Here's what I came up with (running test suite now):

```scala
val tpe1     = extrapolation extrapolate tpe

val (tparamsInitial, tparamsRemaining) = tparams.partition(tpe1.contains)
val tparamClosure = mutable.ListBuffer.empty[Symbol] ++ tparamsInitial
val tparamsTodo   = mutable.ListBuffer.empty[Symbol] ++ tparamsRemaining

var oldLen = 0
while (tparamsTodo.nonEmpty && oldLen != tparamClosure.length) {
  oldLen = tparamClosure.length
  tparamsTodo.filterInPlace { paramTodo =>
    // If our closure is not yet complete (there's a tparam in tparamsClosed whose info refers to a tparam in paramTodo),
    // add `paramTodo` to `tparamClosure`, and drop it from `tparamsTodo`
    !tparamClosure.exists(_.info contains paramTodo) || { 
       tparamClosure += paramTodo; false
    }
  }
}

newExistentialType(tparamClosure.toList, tpe1)
```

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4132,7 +4132,8 @@ trait Types
         tparams0 = tparams1
         tparams1 = tparams filter { p =>
-          tparams1 exists { p1 => p1 == p || (p1.info contains p) }
+          tparams1.exists( (p1: Symbol) => p1 eq p) ||
```
I wonder if it would make sense to use a Set of Symbols and a Set of Types (`p1.map(_.info)`)? How often are the `p1.info` the same and we keep testing against them?

diesalbla replied:
What I am trying to do is to have another "border" collection, that in each iteration keeps the elements found in the latest iteration, and run the exists only on that. This would avoid repeating calls `p1.info.contains(p)` twice for the same pair of `p1, p`.

adriaanm commented:
On a meta-note, I don't mean to take this PR too far off its course, and I certainly don't mean to bikeshed. I do think it's worth poking around here a bit before we move to the next one :-)

diesalbla committed:
Attend Code Review Comments.

This incorporates some changes, suggested by Adrian Moors, to use more
efficient data-structures (in-place), with some modifications.
We use two additional temporary lists, `tparamsBord1` and `bord2`, to
keep a border of the transitive closure: the list of elements
introduced by the previous iteration. This avoids repeating
`Type.contains` computations between same pairs.

We add to the `Collections` a function to partition a list into two ListBuffers.

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4126,16 +4126,27 @@ trait Types
       val extrapolation = new ExistentialExtrapolation(tparams)
       if (flipVariance) extrapolation.variance = Contravariant
       val tpe1     = extrapolation extrapolate tpe
```
I just set a breakpoint here in the debugger to see what sort of data arrives.

I noticed this is called with `NoType`, we could add a fast path for that.

It would also be worthwhile adding a fast path for `tparams.size == 1`, which is pretty common and doesn't need the `ListBuffer` allocations for the general-arity algorithm.

diesalbla replied:
I have added a shortcut for both cases.

retronym commented:
Regarding optimization of `Type.contains`, we could add a fast path for `this: ClassNoArgsTypeRef` (either with an override in that class or with a type test in `Type.contains`), that just checks the `.sym == sym` directly, rather than allocating the `ContainsCollector`. We seem to commonly to ask `typeOf[Nothing].contains(sym)` during type inference.

adriaanm commented:
Looking good, @diesalbla -- thanks! Could you please integrate your commit message as comments into the code, and update the one I wrote originally, to document the purpose of the borders? This stuff if pretty subtle.

I'd also encourage you to incorporate @retronym's suggestions.

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2695,7 +2695,9 @@ trait Types
   private final class AbstractNoArgsTypeRef(pre: Type, sym: Symbol) extends NoArgsTypeRef(pre, sym) with AbstractTypeRef
-  private final class ClassNoArgsTypeRef(pre: Type, sym: Symbol) extends NoArgsTypeRef(pre, sym)
+  private final class ClassNoArgsTypeRef(pre: Type, sym: Symbol) extends NoArgsTypeRef(pre, sym) {
+    override final def contains(sym: Symbol): Boolean = this.sym == sym
```
should could also consider the prefix

diesalbla replied:
Sorry, not an expert on this part.

adriaanm replied:
No prob! This PR is looking awesome.

diesalbla replied:
How would you write it to apply @retronym suggestion? I have reverted that part, to see if it compiles.

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4119,36 +4121,55 @@ trait Types
-    if (tparams.isEmpty) tpe0
+  def existentialAbstraction(tparams: List[Symbol], tpe0: Type, flipVariance: Boolean = false): Type = {
+
+    /* We want to narrow the list of type parameters tparams to only those which are either
```
😍

viktorklang commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4119,25 +4121,55 @@ trait Types
+        val closed    = mutable.ListBuffer.empty[Symbol]
+        var borderAux = mutable.ListBuffer.empty[Symbol]
+        while (border.nonEmpty) {
+          borderAux.clear
```
You could put this at the end of the loop, since it won't be needed to clear it initially.

viktorklang replied:
Looking at it, you could do `swap.clear` after `val swap = border`

diesalbla replied:
I prefer to leave it at the start, to make it clear that `borderAux` is empty. Note that the use of a mutable  `borderAux` is done to save memory: we could just as well declare  `val borderAux = ListBuffer.empty`, which would create a new `ListBuffer` at each iteration.

diesalbla committed:
Add short-cut for Type.contains implementation

diesalbla committed:
existentialAbstraction: add shortcases and document code

Following suggestions and observations by Jason Zaugg:
- We add a shortcuts for a frequent case, of using `NoType`
- We add a shortcut for the case when there is one param.

We also add some comments to describe the transitiveClosure algorithm.

diesalbla committed:
Small details.

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2695,7 +2695,9 @@ trait Types
   private final class AbstractNoArgsTypeRef(pre: Type, sym: Symbol) extends NoArgsTypeRef(pre, sym) with AbstractTypeRef
-  private final class ClassNoArgsTypeRef(pre: Type, sym: Symbol) extends NoArgsTypeRef(pre, sym)
+  private final class ClassNoArgsTypeRef(pre: Type, sym: Symbol) extends NoArgsTypeRef(pre, sym){
+    override def contains(sym: Symbol) = this.sym == sym || super.contains(sym)
```
I'd replace `super.contains(sym)` with `pre.contains(sym)`. Sorry for not being more explicit last time.

adriaanm replied:
After that, good to go!

diesalbla replied:
Ok. Since we are at it, can we apply this trick for other classes too?

diesalbla replied:
It seems to me that specializing the `contains` methods (and maybe `find`) too, so as to avoid creating the `ContainsCollector` can be done in a different PR.

adriaanm replied:
Good question! For example, it would not work for e.g. an alias type ref, because `contains` also checks the right hand side of the alias (by `normalize`ing the type). That makes me wonder, actually, if my previous comment was correct. If the `ClassNoArgsTypeRef` refers to a polymorphic class, `normalize` would turn it into a `PolyType`, but we're ok with our shortcut, since `contains` should not match the type params (because they are local -- there should be no way an outside symbol can be equal), nor their info (a type param's bound could potentially contain a type we're looking for, but we ignore bounds in TypeCollector as far as I can tell). 😅

adriaanm replied:
In any case, yes, let's defer further tweaks to another PR -- this one is more than polished enough to merge.

diesalbla committed:
Type.contains: small optimisation

Directly recur on the `pre` type, before creating a `ContainsCollector`.
